README.md
# SOCS .ics Calendar Parser
A PHP Script that parses an ICS file from SOCS, in a readable format to display elsewhere.

## Features
 * Super easy to use.
 * Once you've included it in your composer package, just one line of code will get you a PHP array of Event Objects.
 * Compatible with Wordpress transients, if you use Wordpress.
 * Compatible with Laravel Caching, if you use Laravel.
 * Actively managed - if you have a feature or idea request, please let me know, or even better - submit a [pull request](https://github.com/fredbradley/socs-calendar-ics-parser/pulls).

## Requirements
This project requires PHP 7. I'm too lazy to write for PHP5.6 these days. Get with the program! 

## How to install
```
composer install fredbradley/socs-calendar-ics-parser
```
## Examples
### Wordpress Widget Example
This is written to work with Wordpress and even has Wordpress Caching (`set_transient`) functionality built in, to speed up page load times. 

If you want to see an example of a widget that uses this code, [please see this piece of code](https://github.com/cranleighschool/cranleigh-socs/blob/master/src/Widgets/UpcomingCalendarEventsWidget.php) in my 'cranleigh-socs' plugin. If you like you can install the entire [Cranleigh SOCS Plugin](https://github.com/cranleighschool/cranleigh-socs), but be aware it's written in Bootstrap 3 with HTML markup specifically for our sites. 

![Wordpress Example](https://user-images.githubusercontent.com/1639226/43061908-eca937b6-8e4e-11e8-8bf7-bd2ea0b2b6b5.png)

If you want some help customising it, I'd be happy to help. 


## Usage
### Options
  * `minNumEvents` (default: 5)
  * `ignoreCache` (default: false) Set to `true` if you don't want get the object from the cache. 
  * `cacheName` (default: 'calendar-cache') if you have more than one instance of this, you should probably give each instance a unique cache name, so they don't overwrite each other
  * `weeksAhead` (default: 15) how many weeks ahead to you want to get the events for
  
  ### Basic Example with Default Options
  ```
  $ics_url = "<YOUR URL FOR YOUR ICS HERE>";
  $calendar = new \FredBradley\SOCSICSParser\CalendarEvents( $ics_url );
  ```
  ### Example with custom options
  ```
  $ics_url = "<YOUR URL FOR YOUR ICS HERE>";
  $options = [
    'minNumEvents' => 10,
    'cacheName' => 'my-music-calendar'
  ];
  $calendar = new \FredBradley\SOCSICSParser\CalendarEvents( $ics_url, $options );
  ```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.0.0](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.15...2.0.0) - 2021-09-30

### Commits

- Version 2. Minimum PHP is now PHP8.0. Created custom Exception for clients to use [`efa3cd1`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/efa3cd16f003352425e075b29cc180fc4f5a6980)
- Update Changelog [`ddb1c65`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/ddb1c65e230e7a20fe6787860801facb8e04b754)

## [1.0.15](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.14...1.0.15) - 2021-09-30

### Commits

- upgrade deps [`1e46ff6`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/1e46ff632f449ffd9a1f9112c83927f6d0e7767d)

## [1.0.14](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.13...1.0.14) - 2020-07-06

### Commits

- fix error [`9623112`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/9623112b05f46b6baec78171e92cb4752a9f41fa)

## [1.0.13](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.12...1.0.13) - 2020-06-22

### Commits

- Update Dependencies [`db9b5b4`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/db9b5b47ba6ec00509183fabe4f4f81404365170)
- Update Dependences [`28a69bc`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/28a69bc7b4b19c8f73f14dafec35bf7c762f5489)
- Update Changelog [`5e588f9`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/5e588f9b63d3d68afd05cbf46d67aefc15f20612)

## [1.0.12](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.11...1.0.12) - 2019-08-07

### Commits

- decode html entities on Location string as well [`01dc5ff`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/01dc5ffbbfa26ab0372435bac9387dcdbe48ad0b)

## [1.0.11](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.10...1.0.11) - 2019-06-14

### Commits

- decode html entities [`f0c8329`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/f0c8329c49d22139123e4b624c04f87cf06e42dd)
- Update Changelog [`5fab384`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/5fab3847067a27c15641a440042c0fb6074827f1)

## [1.0.10](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.9...1.0.10) - 2019-04-12

### Commits

- Updated to show 12 hour clock [`bcac041`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/bcac041d62df5edc74cb1017ac8e93c8ee307a97)

## [1.0.9](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.8...1.0.9) - 2018-10-23

### Commits

- Update Changelog [`163053f`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/163053f25cf4fb927783eaa5ed10dc7a827f30cb)
- Update Version [`2d6e48a`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/2d6e48a2dc53bb6ce2cfed499c6c2f812947e69c)

## [1.0.8](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.7...1.0.8) - 2018-10-23

### Commits

- Update Changelog [`fdf23f9`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/fdf23f915ee84caabc42188370224a1c14c6cba9)
- Update package [`089d82d`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/089d82dc80c0717542721258f7d0995398827e7e)
- Fix laravel cache time [`29f067c`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/29f067c5f2c3460820f96ff6d5e8eddd8b4b65d9)

## [1.0.7](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.6...1.0.7) - 2018-08-07

### Commits

- Update Changelog [`952ffd3`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/952ffd3ba6b7f1232c50e2b9679603635990b737)
- Also show uri query [`f1ae580`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/f1ae58044683087e108c29d41ca96b8333faa167)

## [1.0.6](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.5...1.0.6) - 2018-08-07

### Merged

- Sanitize icsUri [`#1`](https://github.com/fredbradley/socs-calendar-ics-parser/pull/1)

### Commits

- The pull request expected from TJB. [`aa877f1`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/aa877f1bd62fc334ba8da0815830e91ea4010924)
- Update Changelog [`8f59361`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/8f59361ee8932d277b0ef28cf9097ba8be067dc7)
- Update Changelog [`dfe172f`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/dfe172f490e6d8a3771a8f96e66a86b150ab7a09)

## [1.0.5](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.4...1.0.5) - 2018-07-27

### Commits

- Fixes issue with Laravel Cache returning `null` instead of `false` [`f624469`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/f624469a3a549c7131e7a7b30b8a06ebb1bb0c2b)
- Update Changelog [`6a0a702`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/6a0a702e4e42f3197a60a3f6a674d03205117770)
- Update Changelog [`79ff3fa`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/79ff3fa9877f967a0a73d45d41b072dc543cc253)

## [1.0.4](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.3...1.0.4) - 2018-07-25

### Commits

- Version Bump [`fb20936`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/fb20936c5ee0f75876921035285747db0a0eca8f)

## [1.0.3](https://github.com/fredbradley/socs-calendar-ics-parser/compare/1.0.2...1.0.3) - 2018-07-25

### Commits

- Add laravel caching [`a08a691`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/a08a691c577c95348a37c8563b752f4145511dd7)
- Update Changelog [`2ddd43f`](https://github.com/fredbradley/socs-calendar-ics-parser/commit/2ddd43f58d489fd18b2297ff9ef01e01f6a9625f)

src/CalendarEvents.php
<?php
/**
 * Created by PhpStorm.
 * User: fredbradley
 * Date: 19/07/2018
 * Time: 16:04
 */

namespace FredBradley\SOCSICSParser;

use Carbon\Carbon;
use FredBradley\SOCSICSParser\Exceptions\FailedToLoadCalendar;
use FredBradley\SOCSICSParser\Traits\Cache;
use ICal\Event;
use ICal\ICal;

/**
 * Class CalendarEvents
 *
 * @package FredBradley\SOCS
 */
class CalendarEvents
{
    use Cache;
    /**
     * Constants for Time counting. Idea taken from Wordpress!
     */
    const MINUTE_IN_SECONDS = 60;

    const HOUR_IN_SECONDS = 60 * self::MINUTE_IN_SECONDS;

    const DAY_IN_SECONDS = 24 * self::HOUR_IN_SECONDS;

    const LARAVEL = 'laravel';

    const WORDPRESS = 'wordpress';
    /**
     * @var int
     */
    public int $minNumEvents = 5;

    /**
     * @var array
     */
    public array $events = [];
    /**
     * @var bool
     */
    public bool $ignoreCache = true;
    /**
     * @var array
     */
    public array $options = [
        'minNumEvents' => 5,
        'weeksAhead' => 15,
        'ignoreCache' => false,
        'cacheName' => 'calendar-cache',
    ];

    /**
     * @var string
     */
    protected string $icsUri='';

    /**
     * @var array
     */
    private array $unsetVars = [
        'x_microsoft_cdo_alldayevent_array',
        'x_microsoft_cdo_alldayevent',
        'categories_array',
        'summary_array',
        'class_array',
        'location_array',
        'uid_array',
        'uid',
        'attendee',
        'organizer',
        'status',
        'lastmodified',
        'created',
        'transp',
        'sequence',
        'duration',
        'class',
        'dtstamp_array',
        'dtend_array',
        'dtstart_array',
    ];

    /**
     * @var int
     */
    private int $weeksAhead = 15;

    /**
     * @var string
     */
    private string $siteType = '';

    /**
     * @var string
     */
    private string $cacheName = '';

    /**
     * @param  string  $icsUri
     * @param  array  $options
     *
     * @throws \FredBradley\SOCSICSParser\Exceptions\FailedToLoadCalendar
     */
    public function __construct(string $icsUri, array $options = [])
    {
        $this->icsUri = $this->sanitizeWebcalUri($icsUri);
        $this->options = array_merge($this->options, $options);
        foreach ($this->options as $key => $value) {
            $this->{$key} = $value;
        }

        $this->sniffAndSetSiteType();

        $this->loadEvents();
    }

    /**
     * Sanitizer. Our icsUri is expected to be a http or https request.
     *
     * @param string $url
     * @return string (A url)
     */
    private function sanitizeWebcalUri(string $url): string
    {
        $uri = parse_url($url);
        if (in_array($uri[ 'scheme' ], ['http', 'https'])) {
            return $url;
        } else {
            $warning = sprintf("Your .ics file should be a 'http' or 'https' scheme. Not the %s which is what you have set. Please change this. We have changed it to 'http' for you.",
                $uri[ 'scheme' ]);
            error_log("Calendar Warning: " . $warning);
            $scheme = str_replace($uri[ 'scheme' ], "http", $uri[ 'scheme' ]);
            return $scheme . "://" . $uri[ 'host' ] . $uri[ 'path' ] . "?" . $uri[ 'query' ];
        }
    }

    /**
     * Tries to work out what type of system you're running on, to offer the best cacheing experience.
     *
     * @return void
     */
    private function sniffAndSetSiteType(): void
    {
        if (function_exists('add_action') && function_exists('set_transient') && defined('WPINC')) {
            $this->siteType = self::WORDPRESS;
        } elseif (class_exists('\Illuminate\Support\Facades\Cache')) {
            $this->siteType = self::LARAVEL;
        } else {
            $this->siteType = 'unknown';
        }
    }

    /**
     * This function loads the events into $this->events.
     * Depending on settings, it will try and load from the Cache first,
     * to save on page load speed.
     *
     * @return void
     * @throws \FredBradley\SOCSICSParser\Exceptions\FailedToLoadCalendar
     */
    public function loadEvents(): void
    {
        if ($this->fromCache() !== false && $this->ignoreCache === false) {
            $this->events = $this->fromCache();
        } else {
            $calendar = $this->loadCalendar();
            $this->events = $calendar->eventsFromRange(now(), now()->addWeeks($this->weeksAhead));

            $this->manipulateEventsObject();
            $this->saveCache($this->events);
        }
    }

    /**
     * Loads the Calendar into a PHP Object.
     *
     * If it fails, then depending on the site type,
     * it returns an error exception.
     *
     * @return ICal
     * @throws \FredBradley\SOCSICSParser\Exceptions\FailedToLoadCalendar
     */
    public function loadCalendar()
    {
        try {
            $iCal = new ICal($this->icsUri, [
                'defaultSpan' => 1,
                'defaultTimeZone' => 'UTC',
                'defaultWeekStart' => 'MO',
                'defaultCharacterReplacement' => false,
                'skipRecurrence' => false,
                'useTimeZoneWithRRules' => false,
            ]);
            return $iCal;
        } catch (\Exception $e) {
            throw new FailedToLoadCalendar($e->getMessage(), $e->getCode());
        }
    }

    /**
     * Takes the $this->events object and makes it a bit tidier.
     * Removes some variables in the object that we don't need.
     * Adds some variables which make reading the object easier.
     *
     * @return void
     */
    private function manipulateEventsObject(): void
    {
        foreach ($this->events as $event):
            $event->allDayEvent = $this->isAllDayEvent($event);
            $event->multiDayEvent = $this->isMoreThanOneday($event);
            $event->categories = $this->setCategoriesArray($event);
            $event->timeLabel = $this->timeLabel($event);
            $event->calendarName = $this->cacheName;
            $event->summary = html_entity_decode(html_entity_decode($event->summary));
            $event->location = html_entity_decode(html_entity_decode($event->location));

            foreach ($this->unsetVars as $var):
                unset($event->{$var});
            endforeach;
        endforeach;
    }

    /**
     * @param \ICal\Event $event
     *
     * @return bool
     */
    private function isAllDayEvent(Event $event): bool
    {
        if ($event->x_microsoft_cdo_alldayevent === "TRUE") {
            return true;
        } else {
            return false;
        }
    }

    /**
     * @param \ICal\Event $event
     *
     * @return bool
     */
    private function isMoreThanOneday(Event $event): bool
    {
        $start_ts = $event->dtstart_array[ 2 ];
        $end_ts = $event->dtend_array[ 2 ];

        if (($end_ts - $start_ts) > self::DAY_IN_SECONDS) {
            return true;
        } else {
            return false;
        }
    }

    /**
     * @param \ICal\Event $event
     *
     * @return array
     */
    private function setCategoriesArray(Event $event): array
    {
        return explode(", ", $event->categories);
    }

    /**
     * @param \ICal\Event $event
     *
     * @return string
     */
    private function timeLabel(Event $event): string
    {
        $event_start_date = new \DateTime($event->dtstart_tz);
        $event_end_date = new \DateTime($event->dtend_tz);

        if ($this->isAllDayEvent($event)) {
            $time_label = "All Day";
            if (($event_start_date->format("Y-m-d") !== $event_end_date->format("Y-m-d")) && $this->isMoreThanOneDay($event)) {
                $time_label = "Multiday Event. Ends: " . $event_end_date->format("jS M");
            }
        } elseif ($event->dtstart === $event->dtend) {
            $time_label = $event_start_date->format("g:ia");
        } else {
            $time_label = $event_start_date->format("g:ia") . " - " . $event_end_date->format("g:ia");
        }

        return '<i class="fa fa-fw fa-clock-o"></i>' . $time_label;
    }

    /**
     * Depending on $siteType, this will save the input into the appropriate cache.
     *
     * @param array $input
     *
     * @return bool
     */
    private function saveCache(array $input): bool
    {
        if ($this->siteType === 'wordpress') {
            delete_transient($this->cacheName);

            return set_transient($this->cacheName, $input, 6 * self::HOUR_IN_SECONDS);
        }

        if ($this->siteType === 'laravel') {
            return \Illuminate\Support\Facades\Cache::put($this->cacheName, $input, 60);
        }

        return false;
    }

}

